### ISSUE TITLE
Garbage output due to uninitialized data

### ISSUE THREAD

**Issue by alexey-lysiuk** [kcat/openal-soft#448]
While looking at [these](https://forum.zdoom.org/viewtopic.php?t=68623) [issues](https://forum.zdoom.org/viewtopic.php?t=67358), I found a way to reproduce them reliably.
Bisecting commits led to 3007fbf5e52145f002d15f3c8ff4943d992cea98 in which the problem appeared first.

It boils down to the last element of `Device->SourceData` array, which can be accessed but never initialized.
The issue requires a particular sound to be mixed, and it pollutes data buffer with that bogus value [here](https://github.com/kcat/openal-soft/blob/31791c9997d06a3a315d2b936515f643430cd5e9/alc/voice.cpp#L700-L729).
Previously, a whole array was filled with zeroes. With this commit, the last element was skipped.
The [current](https://github.com/kcat/openal-soft/commit/31791c9997d06a3a315d2b936515f643430cd5e9) master doesn't have array fill part, the rest of the code is still there.

Steps to reproduce:

1. Load [Enjay's Waterlab](http://www.aspectsweb.co.uk/enjay/doom/njwlgzd.pk3) with the recent version of GZDoom
2. Use one of the following console commands
```
playsound zombiefemale/sight2
playsound zombiefemale/sight3
```
3. It may produce noticeable clicks at the end of these sounds

To make this always happen, assign some large number to `Device->SourceData[1071]` right after initialization.
Sounds have 96 kHz sampling rate, and this distinguishes them from the rest.
Device has 48 kHz sample rate, and it seems to be related to the issue as well.

Also, I managed to replicate it with `alplay` using the attached [dummy file](https://github.com/kcat/openal-soft/files/4846428/test.wav.zip).

**Comment by kcat** [comment]
Hmm. I can't see where the issue arises. `Device->SourceData[1071]` should get filled as needed. If I set `SourceData[1071] = 4.0f;` in device initialization and don't play those sounds, I don't hear any clicks and it doesn't get overwritten, as would be expected (a mix doesn't need that many source samples to fill the output). If I play those sounds, I still don't hear any clicks and it is being overwritten (due to the high source rate needing a full buffer, it ends up as -0.007874 after playing `zombiefemale/sight3`).

What playback sample rate does the device have (check `snd_status` in the console)? Are you only hearing this on macOS, and not Windows or Linux? The behavior your describing makes it seem like there's an off-by-one error somewhere, that only gets triggered in certain situations, likely relating to the source sample increment/step and device update size (and while the CoreAudio backend unfortunately doesn't handle or report the output device's real update size, since I don't know how to get it, it should still work regardless of the amount of samples it asks from `aluMixData`). I'm having trouble seeing where that error would be, though.

If you're able to add a trace/`ERR()` log call near the affected code to print related parameters (`SrcBufferSize`, `DstBufferSize`, `increment, `DataPosFrac`, etc), and filter it to only show the problem sounds (e.g. when `increment > FRACTIONONE`), that might help pin-point the issue.

**Comment by alexey-lysiuk** [comment]
Yes, that’s on macOS. Device sample rate is 48k regardless of `AL_FREQUENCY` hint.
On Windows, it’s 44100 though, and there is no such issue here.

`SrcBufferSize` was 1071, which seems to be the problem.
`increment` was 8192, and I don’t remember the rest at the moment.

I’ll try other suggestions when I can.

**Comment by kcat** [comment]
I think I found the issue. `SrcDataBuffer` is calculated as the number of source samples needed for the mix:
```
/* Calculate the last written dst sample pos. */
uint64_t DataSize64{DstBufferSize - 1};
/* Calculate the last read src sample pos. */
DataSize64 = (DataSize64*increment + DataPosFrac) >> FRACTIONBITS;
/* +1 to get the src sample count, include padding. */
DataSize64 += 1 + MAX_RESAMPLER_PADDING;
```
When `DstBufferSize==512` and `increment==8192` (`2.0` in fixed-point), the `SrcBufferSize` becomes `1071`, so it'll fill `0...1070` inclusive. That is all that's needed to do the actual mix since the 511th output sample can only use up to the 1070th source sample. However, later on it stores the last-used and some future samples for later:
```
std::copy_n(&SrcData[(increment*DstBufferSize + DataPosFrac)>>FRACTIONBITS],
    chandata.mPrevSamples.size(), chandata.mPrevSamples.begin());
```
Given `DstBufferSize==512` and `increment==2.0` (in fixed-point), storing the 48 samples back results in reading 1024...1071 inclusive. That last increment causes it to "step over" a sample it didn't actually need for the mix, but wants to store for next time.

As long as `increment <= 1.0`, this isn't a problem. The `+1` for the source count makes up for the `-1` on the dest size (and compensates for a partial read of the last sample). But once `increment > 1.0`, the `+1` only ensures it has enough for this mix and not necessarily enough to store for later, which becomes a problem when the sound stops and it uses those missed samples.

I've pushed commit 775187788ad6902feafa6ed07e65e2f9ce50e92e which should address this, presuming that is the actual problem here.
